### PR TITLE
feat(Teleport): add dash teleport resolves #203

### LINKED ISSUE #203: Possible to add a Dash "teleport" script to the toolkit?
The toolkit is already got all kinds of cool stuff, I've been working on a simple game where you can dash in any direction you are facing in 3D Space, so not just along the ground plane. Both would of course be useful in different cases though.

Jan posted a start on slack it looks like, but it seems to mess up the other VRTK scripts at the moment.

Thanks again keep up the good work!

-Ethan

### PR DESCRIPTION
**Pull request by mbbmbbmm**
Dashing Teleport inspired by the implementation of realities 
(realities.io). When dashing through objects their renderers 
can be turned off.
Needs some fine-tuning for the CapsuleCast.
An example scene has been added to show it in action.

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: d2bddba5a131eb7fac168306531017ca42cfb4ac
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/thestonefox/SteamVR_Unity_Toolkit

---
This message was auto-generated by https://gitcop.com

**Comment by thestonefox** [comment]
I'll try and review this PR over this weekend, sorry it's taken so long :(

**Comment by mbbmbbmm** [comment]
Nice! Yeah no problem, I know you're quite busy :)

**Comment by thestonefox** [comment]
Ok, so i finally tried it! :)

just a couple of points/questions/things to clarify:

 * What is the point of the registering the colliders (like with the Renderer Off On Dash script)? If that script is disabled, you can still dash through objects (if you have the player presence script on giving your camera rig a rigid body). It also seems to cause issues by where objects in the game world disappear but dont come back (for example if you dash from behind the green wall on top of the steps, the steps disappear for me).
 * Also, if the renderer Off/On dash script is something that you would probably want to use in a game (for marking things as being collidable/not collidable for instance) then I'd make this a core script and not an example script (if you think it's that generic and reusable)
 * It may also be worth in the readme explaining how you can use the lerp time and min speed to control things like going slow over a long distance but faster over a short distance.
* Also, in the readme, mention the example scene like the other readme bits do.
 *Also! The example scene number is going to need to change as it now clashes :/

Lemme know what you think of the above points, hopefully we can polish this off and have this added to the toolkit soon!

**Comment by mbbmbbmm** [comment]
Thanks for having a look! You are right about the documentation. It was a little late apparently, I also spotted two 'print's I left in the code. I will update the documentation and apply any fixes tomorrow as I am not near my Vive today.

Ok so my thought was to keep it flexible and leave it to the Toolkit-user how to deal with dashing through objects, but at the same time provide a nice starting point for that.
That's why I introduced the CapsuleCast to find the objects that will be hit and the OnDashThrough/OnDashedThrough- events for broadcasting them.
The Toolkit-user might want to do nothing or change the material or make every traveled object glow afterwards or turn the renderers off and so on. Turning the renderer off on dash is just what I would imagine to be the most common scenario which is why I used it for the example. If you think it is better as a core script I can of course move it.  
Now, if in your tests some of the renderers didn't turn back on then something clearly is not right. It could mean that the lerp is never completed and thus the OnDashedThrough-event is not fired. I think the ending condition for the while loop may be poorly chosen. Or I have screwed up something else:)
 
I have to go now, but will have a go tomorrow and also try to answer any further questions later.

**Comment by thestonefox** [comment]
@mbbmbbmm Yeah your idea for making it flexible is a good idea. I'm just not sure if it's working as intended.

I could teleport through objects (using the bezier pointer to teleport over the green wall)

**Comment by mbbmbbmm** [comment]
Do you mean that their renderers remained visible when you were going through them? I should play around some more with the CapsuleCast tomorrow. It's kind of hard to debug, maybe I can make some kind of visualisation to see what is going on.

**Comment by thestonefox** [comment]
Is it not supposed to stop you from teleporting through them? Or does it just turn the renderer off so you don't see yourself moving through the object?

My initial thought was I shouldn't be able to get to the other side of the wall because it was in my way

**Comment by mbbmbbmm** [comment]
It is not supposed to stop you with the current implementation, no! Though that would be rather easy to add as an option. 
Exactly like you said - it should just turn off the renderer, because (to me) it feels much better. I thought if one uses the Bezier pointer one _wants_ to go through objects. With the simple pointer the avoidance of objects is kind of built in.

**Comment by thestonefox** [comment]
If it's not supposed to then it's not broken on that :)

However I did have it when the rendered objects wouldn't be displayed after teleport.

I'd also think an option to stop physically passing through objects would be a good addition

**Comment by mbbmbbmm** [comment]
Ok, so let me see what I can do tomorrow :)

**Comment by joshuha** [comment]
Yeah a collision option would definitely by helpful. A think a big reason people want a dash style teleport is to keep you in the real world space so you can still be hit by other objects and being able to still go through walls or obstacles might defeat that use case.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: b534f0dda2247e0cbbf2b385d03cde8f1b58fff6
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/thestonefox/SteamVR_Unity_Toolkit

---
This message was auto-generated by https://gitcop.com

**Comment by thestonefox** [comment]
You'll want to squash your commits so there is only one commit in the PR. All of the commits around the single feature should be in 1 commit.

If there are distinct sections for an overall feature then it's ok to have separate commits.

**Comment by mbbmbbmm** [comment]
Is there a recommended way to do this? I found this: 
https://makandracards.com/makandra/527-squash-several-git-commits-into-a-single-commit
but I am not sure which method would be best?

**Comment by thestonefox** [comment]
Yeah that link looks about right

**Comment by thestonefox** [comment]
@mbbmbbmm what's the status with the dash stuff? I've been away and just checking i've not missed anything i need to have a look at?

**Comment by mbbmbbmm** [comment]
Oh hey sorry, it's been so long already. No you didn't miss anything, I just need to finish it. I will write to you once I do.

**Comment by thestonefox** [comment]
resolves #203

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: b534f0dda2247e0cbbf2b385d03cde8f1b58fff6
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 6df48e7cdd696be6a7d2dc0f2b96876a7f356cd4
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/thestonefox/SteamVR_Unity_Toolkit

---
This message was auto-generated by https://gitcop.com

File: Assets/SteamVR_Unity_Toolkit/Examples/Resources/Scripts/RendererOffOnDash.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
+namespace VRTK
```
Comment (thestonefox) [inline Assets/SteamVR_Unity_Toolkit/Examples/Resources/Scripts/RendererOffOnDash.cs]: Take the example script out of the VRTK namespace, it should just include the VRTK namespace with:
`using VRTK;`

File: Assets/SteamVR_Unity_Toolkit/Examples/Resources/Scripts/RendererOffOnDash.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        }
+
+        private void RendererOff (RaycastHit [] allHits)
+        {
+            GameObject go = this.transform.gameObject;
+            foreach (RaycastHit hit in allHits)
+            {
+                if (hit.collider.gameObject == go)
+                {
+                    //print ("Turned " + go.name.ToString () + " off");
```
Comment (thestonefox) [inline Assets/SteamVR_Unity_Toolkit/Examples/Resources/Scripts/RendererOffOnDash.cs]: delete comment

File: Assets/SteamVR_Unity_Toolkit/Examples/Resources/Scripts/RendererOffOnDash.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                }
+            }
+        }
+
+        private void RendererBackOn (RaycastHit [] allHits)
+        {
+            GameObject go = this.transform.gameObject;
+            if(wasSwitchedOff)
+            {
+/*                print ("Turned " + go.name.ToString () + " back on")*/;
```
Comment (thestonefox) [inline Assets/SteamVR_Unity_Toolkit/Examples/Resources/Scripts/RendererOffOnDash.cs]: delete comment

File: Assets/SteamVR_Unity_Toolkit/Examples/Resources/Scripts/RendererOffOnDash.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
+namespace VRTK
+{
+    using UnityEngine;
+    using System.Collections;
+    using System;
+
+    public class RendererOffOnDash : MonoBehaviour
+    {
+        bool wasSwitchedOff = false;
+        void Start ()
```
Comment (thestonefox) [inline Assets/SteamVR_Unity_Toolkit/Examples/Resources/Scripts/RendererOffOnDash.cs]: add `private` to method

File: Assets/SteamVR_Unity_Toolkit/Examples/Resources/Scripts/RendererOffOnDash.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        {
+            VRTK_DashTeleport.OnWillDashThruObjects += RendererOff;
+            VRTK_DashTeleport.OnDashedThruObjects += RendererBackOn;
+        }
+        void OnDisable ()
```
Comment (thestonefox) [inline Assets/SteamVR_Unity_Toolkit/Examples/Resources/Scripts/RendererOffOnDash.cs]: add `private` to method

File: Assets/SteamVR_Unity_Toolkit/Scripts/VRTK_DashTeleport.cs [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            Vector3 eyeCameraPositionOnGround = new Vector3(eyeCameraPosition.x, this.transform.position.y, eyeCameraPosition.z);
+            Vector3 eyeCameraRelativeToRig = eyeCameraPosition - transform.position;
+            Vector3 targetEyeCameraPosition = targetPosition + eyeCameraRelativeToRig;
+            Vector3 direction = (targetEyeCameraPosition - eyeCameraPosition).normalized;
+            Vector3 bottomPoint = eyeCameraPositionOnGround + (Vector3.up * capsuleBottomOffset) + direction;
+            Vector3 topPoint = eyeCameraPosition + (Vector3.up * capsuleTopOffset) + direction;
+            float maxDistance = Vector3.Distance(transform.position, targetPosition - direction * 0.5f);
+            RaycastHit[] allHits = Physics.CapsuleCastAll(bottomPoint, topPoint, capsuleRadius, direction, maxDistance);
+
+            // Dash to first obstacle -- not working yet!
```
Comment (thestonefox) [inline Assets/SteamVR_Unity_Toolkit/Scripts/VRTK_DashTeleport.cs]: delete this stuff if it's not working, maybe move it to another local development branch so you can keep working on it in the future.

File: Assets/SteamVR_Unity_Toolkit/Scripts/VRTK_DashTeleport.cs [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        public delegate void DashAction(RaycastHit[] hit);
+        public static event DashAction OnWillDashThruObjects;
+        public static event DashAction OnDashedThruObjects;
+        public float normalLerpTime = 0.1f; //100ms for every dash above minDistanceForNormalLerp, value taken from realities (www.realities.io)
+        public float minSpeedMps = 50.0f; //they also clamp to minimum speed 50m/s to avoid sickness.
+        public float capsuleTopOffset = 0.2f;
+        public float capsuleBottomOffset = 0.5f;
+        public float capsuleRadius = 0.5f;
+
+        // Dash to first obstacle -- not working yet!
```
Comment (thestonefox) [inline Assets/SteamVR_Unity_Toolkit/Scripts/VRTK_DashTeleport.cs]: delete this stuff too if it's not going to be included

**Comment by mbbmbbmm** [comment]
done

File: DOCUMENTATION.md [hunk 8]
```diff
@@ -1672,6 +1673,57 @@ This class allows player movement based on grabbing of `VRTK_InteractableObject`
...
+
+The basic principle is to dash for a very short amount of time, to avoid sim sickness. The default value is 100 miliseconds. This value is fixed for all 
+normal and longer distances. When the distances get very short the minimum speed is clamped to 50 mps, so the dash time becomes even shorter.
+This idea is taken from Realities' "ludicrous speed".
+The minimum distance for the fixed time dash is determined by the minSpeed and normalLerpTime values, if you want to always lerp with a fixed mps speed instead, set the normalLerpTime to a high value. 
+Right before the teleport a capsule is cast towards the target and registers all colliders blocking the way. These obstacles are then broadcast in an event so that for example their gameobjects or renderers can be turned off while the dash is in progress.
+
+### Inspector Parameters
+
+  * **Blink Transition Speed:** The fade blink speed on teleport. Probably should best be set to a value of 0 for dashing.
```
Comment (thestonefox) [inline DOCUMENTATION.md]: Make sure you get the most up to date params from the height adjust teleport docs as looks like you're missing a couple (plus the limit to nav mesh no longer exists)

File: DOCUMENTATION.md [hunk 9]
```diff
@@ -1672,6 +1673,57 @@ This class allows player movement based on grabbing of `VRTK_InteractableObject`
 
 ---
 
+## Dash Teleporter (VRTK_DashTeleport)
+
+### Overview
+
+The dash teleporter extends the height adjust teleporter and allows to have the `[CameraRig]` dashing to a new teleport location.
```
Comment (thestonefox) [inline DOCUMENTATION.md]: Just check this doc and make sure you're not putting in unnecessary line endings, you don't need to limit characters per line in this new documentation file :)

**Comment by mbbmbbmm** [comment]
and done. I was wondering if I should get rid of the blink parameters in the doc and point to the height adjust paragraph instead but left them in for completeness.. maybe somebody _does_ want to use blinking while dashing

**Comment by thestonefox** [comment]
I'll look to re-implement your solution, because it's definitely something that I'd like in the toolkit but I assume you're busy at the moment to fix any issues (which there may be many conflicts now!)

**Comment by mbbmbbmm** [comment]
Yes sorry, I am quite busy at the moment indeed. Hopefully october will be a little bit more relaxed :-)
-- I had changed the documentation according to plan, but something was broken - there were conflicts. I asked you on the slack, but you weren't online at the time and it somehow got lost.
We could also do it this weekend if you dare to enter another remote git help session with me ;-)
Whatever works best for you!

**Comment by thestonefox** [comment]
I'm not going to be around this weekend :( :(

But there are probably a few breaking changes now, even the script path has changed!

may be better pulling the latest master and copying the changes over to a new branch forked of the latest master? But if you don't have time, don't worry I can do it, the only problem with that is you don't get your name in lights as a contributor :(

i'm happy to wait though if you want

**Comment by mbbmbbmm** [comment]
Yes, since it is my first contribution to something I kinda want to "do it right" :)
I will do as you suggested with a new fork this weekend - I think should be able to do it.
Should I fail I'll hand it over to you, OK?

**Comment by thestonefox** [comment]
Yeah cool no worries take your time :)

**Pull request closed without merging**